### PR TITLE
Adding queue name restriction support to cassieq

### PR DESCRIPTION
**Pull request by jakeswenson**
Adding queue name restriction support to cassieq signed urls
Fixes issue #82

### CODE REVIEW

File: core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java [hunk 1]
```diff
@@ -93,7 +94,8 @@ public Response generateAuthUrl(@NotNull GetAuthQueryParamsRequest request) thro
                         request.getAccountName(),
                         authorizationLevels,
                         request.getStartTime(),
-                        request.getEndTime());
+                        request.getEndTime(),
```
Comment (jakeswenson) [inline core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java:97]: hey we need a good way to keep this and the query string builder below in sync. Lets think about a way to do this

Comment (jakeswenson) [inline core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java:97]: should probably only have a query string builder, and it computes the signature

Comment (jakeswenson) [inline core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java:97]: that'd be hot

File: core/src/main/java/io/paradoxical/cassieq/discoverable/auth/SignedUrlParameters.java [hunk 2]
```diff
@@ -56,7 +61,9 @@ public boolean verify(final VerificationContext context) {
         if (verified) {
             final Instant now = context.getClock().now();
             return startDateTime.map(now::isAfter).orElse(true) &&
-                   endDateTime.map(now::isBefore).orElse(true);
+                   endDateTime.map(now::isBefore).orElse(true) &&
```
Comment (devshorts) [inline core/src/main/java/io/paradoxical/cassieq/discoverable/auth/SignedUrlParameters.java]: i think we should encapsulate these predicates, as they are starting to grow. like 

```
notExpired(clock) && queueAllowed()
```

File: core/src/test/java/io/paradoxical/cassieq/unittests/ApiTester.java [hunk 3]
```diff
@@ -182,19 +183,67 @@ public void test_query_auth_authenticates() throws IOException, InvalidKeyExcept
...
+        final AdminClient adminClient = AdminClient.createClient(server.getAdminuri().toString());
+
+        final AccountName accountName = AccountName.valueOf("test_query_auth_restricts_queue");
+        final QueueName queueName = QueueName.valueOf("test_query_auth_restricts_queue");
+        final QueueName badQueueName = QueueName.valueOf(queueName.get() + "bad");
+
+        final Response<AccountDefinition> createAccount = adminClient.createAccount(accountName).execute();
+        final AccountKey accountKey = createAccount.body().getKeys().get(WellKnownKeyNames.Primary.getKeyName());
+
+        final CassieqApi fullAccessClient = CassieqApi.createClient(server.getBaseUri(), CassieqCredentials.key(accountName, accountKey));
```
Comment (devshorts) [inline core/src/test/java/io/paradoxical/cassieq/unittests/ApiTester.java:196]: solid naming here

File: core/src/main/java/io/paradoxical/cassieq/discoverable/auth/SignedUrlParameters.java (hunk 2 above)
Comment (jakeswenson) [inline core/src/main/java/io/paradoxical/cassieq/discoverable/auth/SignedUrlParameters.java]: super good idea. :+1:

File: core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java (hunk 1 above)
Comment (devshorts) [inline core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java:97]: lets decouple the data from the action. lets create a signature payload that can be passed to a signer and a query param generator. the query param generaotr can expose like a 

```
var signatureParams = new SigningParameters(
                        request.getAccountName(),
                        authorizationLevels,
                        request.getStartTime(),
                        request.getEndTime(),
                        request.getQueueName());

String queryParams = new QueryParamGenerator(signatureParams, mac)
```

or something

Comment (jakeswenson) [inline core/src/main/java/io/paradoxical/cassieq/admin/resources/api/v1/PermissionsResource.java:97]: interested idea! will file a task an play with it a bit

**Comment by devshorts** [comment]
:monorail:

**Pull request merged**